model/riscv_insts_zkn.sail

```diff
@@ -0,0 +1,412 @@
+  RETIRE_SUCCESS
+}
+
+function clause execute (SHA512SUM0R(rs2, rs1, rd)) = {
+  X(rd) = EXTS((X(rs1) << 25) ^ (X(rs1) << 30) ^ (X(rs1) >> 28) ^
+               (X(rs2) >>  7) ^ (X(rs2) >>  2) ^ (X(rs2) <<  4) );
+  RETIRE_SUCCESS
+}
+
+function clause execute (SHA512SUM1R(rs2, rs1, rd)) = {
+  X(rd) = EXTS((X(rs1) << 23) ^ (X(rs1) >> 14) ^ (X(rs1) >> 18) ^
+               (X(rs2) >>  9) ^ (X(rs2) << 18) ^ (X(rs2) << 14) );
+  RETIRE_SUCCESS
+}
+
+/*
+ * Scalar Cryptography Extension - Scalar 64-bit AES instructions
+ * ----------------------------------------------------------------------
+ */
+
+union clause ast = AES64KS1I : (bits(4), regidx, regidx)
+union clause ast = AES64KS2  : (regidx, regidx, regidx)
+union clause ast = AES64IM   : (regidx, regidx)
+union clause ast = AES64ESM  : (regidx, regidx, regidx)
+union clause ast = AES64ES   : (regidx, regidx, regidx)
+union clause ast = AES64DSM  : (regidx, regidx, regidx)
+union clause ast = AES64DS   : (regidx, regidx, regidx)
+
+mapping clause encdec = AES64KS1I (rcon, rs1, rd) if haveZkne64()
+  <-> 0b00 @ 0b11000 @ 0b1 @ rcon @ rs1 @ 0b001 @ rd @ 0b0010011
+
+mapping clause encdec = AES64IM (rs1, rd) if haveZkne64()
+  <-> 0b00 @ 0b11000 @ 0b00000 @ rs1 @ 0b001 @ rd @ 0b0010011
+
+mapping clause encdec = AES64KS2 (rs2, rs1, rd) if haveZkne64()
+  <-> 0b01 @ 0b11111 @ rs2 @ rs1 @ 0b000 @ rd @ 0b0110011
+
+mapping clause encdec = AES64ESM (rs2, rs1, rd) if haveZkne64()
+  <-> 0b00 @ 0b11011 @ rs2 @ rs1 @ 0b000 @ rd @ 0b0110011
+
+mapping clause encdec = AES64ES (rs2, rs1, rd) if haveZkne64()
+  <-> 0b00 @ 0b11001 @ rs2 @ rs1 @ 0b000 @ rd @ 0b0110011
+
+mapping clause encdec = AES64DSM (rs2, rs1, rd) if haveZkne64()
+  <-> 0b00 @ 0b11111 @ rs2 @ rs1 @ 0b000 @ rd @ 0b0110011
+
+mapping clause encdec = AES64DS (rs2, rs1, rd) if haveZkne64()
+  <-> 0b00 @ 0b11101 @ rs2 @ rs1 @ 0b000 @ rd @ 0b0110011
+
+mapping clause assembly = AES64KS1I (rcon, rs1, rd)
+  <-> "aes64ks1i" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1) ^ sep() ^ hex_bits_4(rcon)
+
+mapping clause assembly = AES64KS2 (rs2, rs1, rd)
+  <-> "aes64ks2" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1) ^ sep() ^ reg_name(rs2)
+
+mapping clause assembly = AES64IM (rs1, rd) 
+  <-> "aes64im" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1)
+
+mapping clause assembly = AES64ESM (rs2, rs1, rd)
+  <-> "aes64esm" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1) ^ sep() ^ reg_name(rs2)
+
+mapping clause assembly = AES64ES (rs2, rs1, rd)
+  <-> "aes64es" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1) ^ sep() ^ reg_name(rs2)
+
+mapping clause assembly = AES64DSM (rs2, rs1, rd)
+  <-> "aes64dsm" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1) ^ sep() ^ reg_name(rs2)
+
+mapping clause assembly = AES64DS (rs2, rs1, rd)
+  <-> "aes64ds" ^ spc() ^ reg_name(rd) ^ sep() ^ reg_name(rs1) ^ sep() ^ reg_name(rs2)
+
+function clause execute (AES64KS1I(rcon, rs1, rd)) = {
+  if(sizeof(xlen)==32) then RETIRE_FAIL else {
```

Comment:
I've cheated / compromised here. This allows it to compile the same source for RV32/64. It _doesn't_ support dynamic XLEN changes, but it does keep the code as readable as possible. Solving how to structure the code for dynamic XLEN changes is clearly a problem to solve, but I don't feel like I can solve it now without making this PR substantially more difficult for people to read and manage.

I hope this is acceptable, but if it's imperative to solve the dynamic XLEN problem now as much as possible, I can do the necessary changes.

Reply:
You should be able to just `assert(sizeof(xlen) != 32);` at the start of these (also has the advantage that it blows up if you messed up the decode predicates, rather than silently being a NOP that doesn't increment minstret; RETIRE_FAIL only indicates whether or not the instruction should increase minstret, it's up to the instruction to actually raise the relevant exception).

Reply:
Oh that's much nicer, I'll do that.